File: medium/129. Sum Root to Leaf Numbers/129. Sum Root to Leaf Numbers.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        # def collect_paths(
        #     tree: Optional[TreeNode], path: str, paths: List[int]
        # ) -> None:
        #     if not tree:
        #         return

        #     path = path + str(tree.val)

        #     if not tree.left and not tree.right:
        #         paths.append(int(path))
        #     else:
        #         collect_paths(tree.left, path, paths)
        #         collect_paths(tree.right, path, paths)

        #     path = path[: len(path) - 1]

        # paths = []
        # collect_paths(root, "", paths)

        # return sum(paths)
        
        def dfs(tree: Optional[TreeNode], curr_number: int) -> int:
            if not tree:
                return 0
            
            curr_number = curr_number * 10 + tree.val
            
            if not tree.right and not tree.left:
                return curr_number
            
            return dfs(tree.left, curr_number) + dfs(tree.right, curr_number)
        
        return dfs(root, 0)


solution = Solution()
# print(solution.sumNumbers(root=[1, 2, 3]))
# print(solution.sumNumbers(root=[4, 9, 0, 5, 1]))
print("123"[:2])

File: medium/129. Sum Root to Leaf Numbers/129. Sum Root to Leaf Numbers.ts
/**
 * Definition for a binary tree node.
 */
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}
function sumNumbers(root: TreeNode | null): number {
    const collectPaths = (tree: TreeNode | null, path: string, paths: number[]) => {
        if (!tree)
            return

        path = path + tree.val

        if (!tree.left && !tree.right)
            paths.push(Number(path))
        else {
            collectPaths(tree.left, path, paths)
            collectPaths(tree.right, path, paths)
        }

        path = path.slice(0, path.length - 1)
    }

    const treePaths: number[] = []
    collectPaths(root, "", treePaths)

    let sum = 0
    for (const res of treePaths) {
        sum += res
    }

    return sum
};

export {}